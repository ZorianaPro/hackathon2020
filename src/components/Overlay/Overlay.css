.Overlay {
    width: 100%;
    height: 100%;
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    background-color: rgba(31, 31, 31, 0.8);
    opacity: 1;
    display: block;
    transition: all ease-in 100ms;
    z-index: 2
}

.Overlay .Overlay-Container {
    position: relative;
    width: 100%;
    height: 100%;
}

.Overlay .Overlay-Box {
    position: absolute;
    top: 5%;
    left: 50%;
    transform: translateX(-50%);
    width: 70%;
    background-color: var(--white-color);
    border-radius: 0.5em;
    min-width: 17em;
    padding: 2em;
    max-width: 45em;
}

.Overlay .Overlay-Close {
    position: absolute;
    right: 2em;
    top: 2em;
    cursor: pointer;
}
