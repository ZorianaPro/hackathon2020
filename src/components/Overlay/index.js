import React, { useState, useCallback, useEffect } from 'react';
import { component } from '@haensl/services';
import XCloseSVG from '../XCloseSVG';
import './Overlay.css';

const Overlay = ({
  animationDurationMs = 200,
  children,
  onClose,
  shouldClose
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const [isOpening, setIsOpening] = useState(true);
  const [isClosing, setIsClosing] = useState(false);

  const closeOverlay = useCallback(() => {
    setIsClosing(true);
  }, []);

  useEffect(() => {
    if (typeof shouldClose !== 'boolean') {
      return;
    }

    if (shouldClose && isOpen) {
      setIsClosing(true);
    } else if (!(shouldClose || isOpen)) {
      setIsOpening(true);
    }
  }, [
    shouldClose,
    isOpen
  ]);

  useEffect(() => {
    if (isOpening) {
      const timerId = setTimeout(() => {
        setIsOpen(true);
        setIsOpening(false);
      }, animationDurationMs);

      return () => {
        clearTimeout(timerId);
      };
    }
  }, [
    isOpening,
    animationDurationMs
  ]);

  useEffect(() => {
    if (isClosing) {
      const timerId = setTimeout(() => {
        setIsOpen(false);
        setIsClosing(false);
        if (typeof onClose === 'function') {
          onClose();
        }
      }, animationDurationMs);

      return () => {
        clearTimeout(timerId);
      };
    }
  }, [
    isClosing,
    animationDurationMs,
    onClose
  ]);

  if (!(isOpen || isOpening || isClosing)) {
    return null;
  }

  const className = component.className({
    isOpen,
    isOpening,
    isClosing
  }, 'Overlay');

  return (
    <div className={ className }>
      <div className="Overlay-Content"
        style={{
          animationDuration: `${animationDurationMs}ms`
        }}>
        <div className="Overlay-Mask"
          onClick={ closeOverlay }
        />
        <div className="Overlay-Box">
          <div className="Overlay-Close"
            onClick={
              closeOverlay
            }>
            <XCloseSVG/>
          </div>
          {
            children
          }
        </div>
      </div>
    </div>
  );
};

export default Overlay;
