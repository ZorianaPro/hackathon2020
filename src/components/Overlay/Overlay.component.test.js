import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import Overlay from './';

describe('Overlay', () => {
  let overlay;

  it('renders without crashing', () => {
    expect(shallow.bind(shallow, <Overlay />))
      .not
      .toThrow();
  });

  it('renders as expected', () => {
    expect(shallow(<Overlay />))
      .toMatchSnapshot();
  });

  describe('children', () => {
    it('renders all the children', () => {
      expect(
        shallow(
          <Overlay>
            <p>test</p>
          </Overlay>
        ).exists('p')
      ).toBe(true);
    });
  });

  describe('when clicking on the mask', () => {
    beforeEach(() => {
      overlay = mount(<Overlay/>);
      overlay.find('.Overlay-Mask')
        .simulate('click');
    });

    it('closes the overlay', () => {
      expect(overlay.exists('.Overlay--isClosing'))
        .toBe(true);
    });
  });

  describe('animationDurationMs', () => {
    beforeAll(() => {
      jest.useFakeTimers();
      overlay = mount(<Overlay animationDurationMs={ 100 } />);
    });

    it('renders an opening overlay', () => {
      expect(overlay.exists('.Overlay--isOpening'))
        .toBe(true);
    });

    describe('after animationDurationMs milliseconds', () => {
      beforeAll(() => {
        act(() => {
          jest.advanceTimersByTime(100);
        });

        overlay.update();
      });

      it('no longer renders an opening overlay', () => {
        expect(overlay.exists('.Overlay--isOpening'))
          .toBe(false);
      });

      it('renders an open overlay', () => {
        expect(overlay.exists('.Overlay--isOpen'))
          .toBe(true);
      });
    });
  });

  describe('shouldClose', () => {
    let onClose;

    beforeAll(() => {
      jest.useFakeTimers();
      onClose = jest.fn();
      overlay = mount(<Overlay onClose={ onClose }/>);
    });

    afterAll(() => {
      jest.useRealTimer();
    });

    it('renders an opening overlay', () => {
      expect(overlay.exists('.Overlay--isOpening'))
        .toBe(true);
    });

    describe('when shouldClose is true', () => {
      beforeAll(() => {
        act(() => {
          jest.runOnlyPendingTimers();
        });

        overlay.update();
      });

      it('renders an open overlay', () => {
        expect(overlay.exists('.Overlay--isOpen'))
          .toBe(true);
      });

      beforeEach(() => {
        act(() => {
          overlay.setProps({ shouldClose: true });
        });

        overlay.update();
      });

      it('starts the closing animation', () => {
        expect(overlay.exists('.Overlay--isClosing'))
          .toBe(true);
      });

      describe('after the animation has finished', () => {
        beforeEach(() => {
          act(() => {
            jest.runOnlyPendingTimers();
          });

          overlay.update();
        });

        it('removes the isOpen class', () => {
          expect(overlay.exists('.Overlay--isOpen'))
            .toBe(false);
        });

        it('removes the isClosing class', () => {
          expect(overlay.exists('.Overlay--isClosing'))
            .toBe(false);
        });

        it('does not render anything', () => {
          expect(overlay.exists('.Overlay > *'))
            .toBe(false);
        });

        it('calls the onClose callback', () => {
          expect(onClose)
            .toHaveBeenCalled();
        });
      });
    });
  });
});
