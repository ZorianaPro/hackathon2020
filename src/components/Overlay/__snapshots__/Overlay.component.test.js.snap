// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay renders as expected 1`] = `
<div
  className="Overlay Overlay--isOpening"
>
  <div
    className="Overlay-Content"
    style={
      Object {
        "animationDuration": "200ms",
      }
    }
  >
    <div
      className="Overlay-Mask"
      onClick={[Function]}
    />
    <div
      className="Overlay-Box"
    >
      <div
        className="Overlay-Close"
        onClick={[Function]}
      >
        <XCloseSVG />
      </div>
    </div>
  </div>
</div>
`;
